### ISSUE TITLE
Scheduled snapshot expiry

### ISSUE THREAD

**Issue by 19wolf** [lxc/lxd#5387]
I saw in [this post](url
https://github.com/lxc/lxd/issues/2610) that there may be interest in snapshot expiry. The thread said 

> If there is interest for this part specifically, please file a separate issue about it and we'll schedule it to be implemented after this one.

So here I am, filing a separate issue about it because I didn't see another one. I don't need hourly snapshots of my container going back into the last week, only the last day. And I only need weekly snapshots into the last 2 weeks, etc.

**Comment by stgraber** [comment]
@monstermunchkin so I think we'll want to expand what we have a bit with:
 - Configuration option for expiration of snapshot (defaults to no expiry)
 - Expiry date stored against the snapshots in DB and API (updatable through API)
 - Background logic to actually expire things, likely same mechanism as is done for backups

Sounds good?

**Comment by monstermunchkin** [comment]
Sounds good

**Comment by monstermunchkin** [comment]
> Expiry date stored against the snapshots in DB and API (updatable through API)

Should the expiry date be added to the `Container` struct in db, or do we need a separate struct (and table)? If it were to be added to `Containers`, we would need to somehow handle the expiry date on normal containers.

For the API part I was thinking of allowing `PUT` on `/1.0/containers/<name>/snapshots/<name>`. The user could then set the expiry via `lxc config edit c1/snap0` (which is invalid at the moment).

@stgraber What do you think?

**Comment by stgraber** [comment]
In the Go API we already use a different structs for snapshots (ContainerSnapshot) so we could add an ExpiresAt field to that, add a DB column for it and keep that only for containers, indeed allowing PUT to update it (and reject all other changes).

Looks like we may want to cleanup that ContainerSnapshot struct a bit too as it may be lacking some fields that were added to containers (and that we forgot to add there) and I spotted CreationDate which should really be CreatedAt to line up with the contianer one.

Luckily it's a sufficiently unused struct that we should be able to fix it up a bit without breaking anyone.
